standard/types.md

```diff
@@ -705,3 +714,182 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
 - `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
 - Any *enum_type*.
 - Any user-defined *struct_type* that is not a constructed type and contains instance fields of *unmanaged_type*s only.
+- In unsafe code ([§23.2](unsafe-code.md#232-unsafe-contexts)), any *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)).
+
+## §Types-And-Nullability Reference Types and nullability
+
+### §Nullable-Types-General General
+
+Any expression of a reference type can either refer to an object, or be the value `null`, which indicates "no reference". The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to permit null values or not. A compiler can provide diagnostics when a expression is not used according to that intent.
+
+There are two forms of nullability for reference types:
+
+- *nullable*: A *nullable-reference-type* can be assigned `null`. Its default null state is *maybe-null*.
+- *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. Its default null state is *not-null*.
+
+Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the same underlying type, `R`; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. An identity conversion exists among a nullable reference type and its corresponding non-nullable reference type (§10.2.2).
+
+> *Note:* By definition, a reference type is nullable; that is, a variable of that type can either contain a reference to an object or be the value `null`, which indicates “no reference.” In a disabled nullable context, reference types aren't annotated. Therefore, they are non-nullable reference types. However, the disabled context means the compiler does not track null state and does not issue nullable related warnings. The disabled nullable context matches the previous standard behavior for reference types. *end note*
```

Comment:
a) Forward reference to *nullable context* should be added

b) This reads wrong to me. The previous standard behaviour was that all reference types were *nullable*, i.e. `T` was nullable and there was no `T?`. In the new system `T` becomes non-nullable and `T?` is nullable.

Let’s try some wordsmithing – as a *starting point only*:

> *Note:* By definition, a reference type is nullable; that is, a variable of that type can either contain a reference to an object or be the value `null`, which indicates “no reference.” In a disabled nullable context, no distinction is made between `T` and `T?`, variables declared either way are treated as nullable, no null state tracking is required, and therefore no consequential warnings need be issued. The disabled nullable context matches the previous standard behavior for reference types. *end note*

This uses “is required” and “need” but it’s not clear to me they are correct. Under the current system I do not believe that an implementation which choose to track the possible null value of reference variables, and issue warnings that a value might be/is null, would be non-compliant.

The new system assists, but does not require, an implementation to track null state and issue implementation-specific warnings – so what does it mean to “disable” this non-compulsory system? Does an implementation which already did its own tracking and warning, or has developed better tracking (say by tracking known null values) now have to disable it if the new optional, but specified, null state is disabled?

So I went with ”no [...] is required” for the tracking, and “no [...] warnings need be issued which does allow tracking and warnings, but would that be non-compliant?

> *“I am a compliant implementation and you have disabled nullability, so with apologies I cannot tell you why this bit of code will fail. If you'd like to know enable nullability checking and then I can tell you.”* some future anthropomorphised implementation ;-) [yes, it's been a long day!]

Maybe “disabled” in this context should mean ”don’t flood me with inaccurate warnings but if you are *certain* you may tell me”? But I haven't spent any time on how we might specify that!

Reply:
Things we're concerned about in 732:

- Null state tracking is always enabled, regardless of context
- Now we have two flags in one context, we shouldn't just talk about the context being disabled
- Nigel's point around "everything was previously nullable" is a natural reaction, and we need to work hard at figuring out how to square that circle.

Might be worth pairing on this.